### ISSUE TITLE
GridMap does not convert collision data and it's not possible to define this data manually

### ISSUE THREAD

**Issue by Giacom** [godotengine/godot#11888]
**Operating system or device, Godot version, GPU Model and driver (if graphics related):**
master branch - b368ce1719746c69828688b9bf88396dd10dfe26
Windows 10
Compiler: MSVC x64
Debug

**Issue description:**
GridMap does not import the collisions of the tiles being converted in the scene.
It will also not let you manually add collision shapes through the inspector.

Trying to resize the array results in this log message:

``` 
scene\resources\mesh_library.cpp:237 - Condition ' p_shapes.size() & 1 ' is true.
Resize Array
```

**Steps to reproduce:**
1. Open the 'Platformer 3D' project.
2. Select the tiles.res file, go through the items in the inspector and notice how it has shapes for collision data.
3. Delete the tiles.res file in the project structure.
4. Open the tiles.scn file.
5. Attempt to recreate the tiles.res file with 'Convert To.. -> MeshLibrary'.
6. Select the new file and notice how no shapes are defined.
7. Try to add shapes manually and notice how it will not let you change the array number.

**Comment by djrm** [comment]
The current implementation has a static body per octant, so as far as i know its not currently possible to have per cell collision data, however the gridmap should then provide a general collision properties that will be applied to every octant.

**Comment by akien-mga** [comment]
We have now entered release freeze for Godot 3.0 and want to focus only on release critical issues for that milestone. Therefore, we're moving this issue to the 3.1 milestone, though a fix may be made available for a 3.0.x maintenance release after it has been tested in the master branch during 3.1 development. If you consider that this issue is critical enough to warrant blocking the 3.0 release until fixed, please comment so that we can assess it more in-depth.

**Comment by Giacom** [comment]
I'd say this issue is pretty important to fix.

**Comment by Nutriz** [comment]
Well, like Giacom said, this one is very important, no collision == no use of GridMap in much case. While Godot 3 is focused a lot on 3D, it's a bad thing not to be able to use it

**Comment by reduz** [comment]
I don't understand this issue. You can't create a MeshLibrary manually, you need to generate it from a scene. What you ask for is an enhancement but it's clearly not a bug.

**Comment by reduz** [comment]
wait i think i misunderstood

**Comment by kubecz3k** [comment]
I was able to reproduce the issue with 531dc2f435c1d64a2e3ef0710cc05a7c13969587.
I also noticed there is no visible rendering material after the generation of the MeshLibrary (GridMap geometry is white)

**Comment by Alastor01** [comment]
This issue is still present in 3.1 Beta 3. Although I am not sure if it is exactly the same...
After conversion, the collision shapes attached to GridMap tiles become deformed and completely useless. Basically all my BoxColliders have been deformed during conversion.
Please look at the difference between Mesh scene collision shape and it's corresponding GridMap Meshlib:
![tile](https://user-images.githubusercontent.com/43552811/52524852-98ff5700-2c99-11e9-8f05-3eb8bf6015e8.PNG)
![tile_gridmap](https://user-images.githubusercontent.com/43552811/52524853-9997ed80-2c99-11e9-8222-aeac6f644755.PNG)

This is important issue which prevents effective use of GridMaps at this stage.

**Comment by ilmoeuro** [comment]
This seems to be fixed in 3.1, I was able to create a non-cube collision shape that retained its dimensions in the game. Note that the collision shape has scale factor of 1, the shape was determined using the "extents" property.

![image](https://user-images.githubusercontent.com/1074146/56467163-1b3b9280-6424-11e9-904a-984765dafe90.png)

![image](https://user-images.githubusercontent.com/1074146/56467168-31495300-6424-11e9-9519-90db789b0c3e.png)

**Comment by akien-mga** [comment]
Thanks, closing.

**Comment by madmiraal** [comment]
Reopening, because I'm still experiencing the issue as [originally described](https://github.com/godotengine/godot/issues/11888#issue-263551699) in both 3.2.2 and master 9e34ba4 i.e. when trying to create a `MeshLibrary` using ```Scene -> Convert To... -> MeshLibrary...``` or add to a `MeshLibrary` using ```Mesh Library -> Import from Scene```  the MeshLibrary doesn't contain the scene's `CollisionShapes`, and there is no way to add them manually.

**Comment by Datoh** [comment]
Any hack to add the `CollisionShape` to the `MeshLibrary` ?

**Comment by Datoh** [comment]
I found the issue in my case. When converting a scene to Mesh Library, to take into account collision shapes, you must have the following tree:
`MeshInstance>StaticBody>CollisionShapes[]`
In my case I had the followinf tree:
`StaticBody>[MeshInstance, CollisionShape]`

ping @madmiraal

**Comment by ElasticCracker** [comment]
I am using Godot v3.4 stable as of today, branch [206ba70f4] and am having the same issue.

I tried using the scene order that @Datoh described without success. I even opened up the gridmap demo project in the Docs (https://docs.godotengine.org/en/stable/tutorials/3d/using_gridmaps.html) 
and have the same issue. Is there a workaround?

**Comment by Klowner** [comment]
Tried to reproduce with both 33b80bd933 (3.x) and 040f49ed6e (4.x), I used the gridmap demo project from the docs (as @Schwarcraft mentioned) and I see that it has the scene aranged in `MeshInstance>StaticBody>CollisionShapes[]` order that @Datoh previously described.

Converting the `MeshLibrary_Source.tscn` to a `MeshLibrary` produces the expected library items and they appear to have collision shapes as one would expect.

Unless I'm missing some crucial detail, I am unable to reproduce this one.
